### PR TITLE
Adds RestartFlow to restart only on failures #24421

### PR DESCRIPTION
**Pull request by nachinius**
Refs #24421

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/RestartSpec.scala [hunk 1]
```diff
@@ -667,6 +675,65 @@ class RestartSpec extends StreamSpec(Map("akka.test.single-expect-default" -> "1
...
+      val (created, source, flowInProbe, flowOutProbe, sink) = setupFlow(shortMinBackoff, shortMaxBackoff, -1, onlyOnFailures)
+
+      source.sendNext("a")
+      flowInProbe.requestNext("a")
+      flowOutProbe.sendNext("b")
+      sink.requestNext("b")
+
+      flowOutProbe.sendNext("complete")
+      sink.request(1)
+      sink.expectComplete()
```
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/RestartSpec.scala]: request should technically not be needed there,

**Review by ktoso — APPROVED** [review]
LGTM :)

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/RestartSpec.scala (hunk 1 above)
Comment (nachinius) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/RestartSpec.scala]: thks for the tip

**Pull request merged**